cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -182,24 +183,30 @@ public ProgramStatus getProgramStatus(ProgramId programId) throws Exception {
       throw new NotFoundException(appId);
     }
 
+    programId = new ProgramId(programId.getNamespace(),
+                              programId.getApplication(),
+                              appSpec.getAppVersion(),
```

Comment:
still have the same question, shouldn't this method take a ProgramReference instead of a ProgramId? Or have one method that takes a ProgramReference and another that takes a ProgramId?

Reply:
Good idea, will update this

Reply:
Actually this was also involved in @jingtaowu3 's open PR in https://github.com/cdapio/cdap/pull/14675

I'll revert the single getLatest call changes in this PR and only touch batch APIs instead. That way this method contract is still correct